File: treble/render/web/src/renderer.ts
// Shared cell-grid renderer (spec §6.1; invariant I6).
//
// Consumes the SAME canonical layout tree the conformance suite compares
// and the TUI renders. Nothing here resolves data or interprets a screen
// definition: if this file needed to know what DES means, the abstraction
// would have failed.
//
// Used by both the Tauri desktop shell and the browser client, which is
// why it lives under treble/render/web rather than inside apps/desktop.

export type Attr =
  | "label" | "editable" | "link" | "positive" | "negative"
  | "warning" | "stale" | "model_derived" | "blink" | "emphasis";

export interface LayoutCell {
  at: [number, number];
  text: string;
  attrs: Attr[];
  link: string | null;
  provenance: string | null;
  input: string | null;
}

export interface LayoutPane {
  region: [number, number, number, number]; // row, col, height, width
  type: string;
  binding: string;
  data: Array<Array<string | number | null>>;
}

export interface LayoutTree {
  mnemonic: string;
  tab: string;
  grid: [number, number];
  stale: boolean;
  cells: LayoutCell[];
  panes: LayoutPane[];
  footnotes: string[];
}

// §6.3 colour semantics. Tokens, never colours, come from the definition;
// this is the desktop theme's mapping of them.
const STYLES: Record<Attr, string> = {
  label: "color:#d78700",
  editable: "color:#ffffff;background:#1c1c1c",
  link: "color:#00afd7;cursor:pointer;text-decoration:underline",
  positive: "color:#00af5f",
  negative: "color:#d70000",
  warning: "color:#d7d700",
  stale: "color:#767676",
  // §5.4: model-derived values carry a dotted underline and expand via SPTR.
  model_derived: "text-decoration:underline dotted",
  blink: "animation:treble-blink 1s step-end infinite",
  emphasis: "font-weight:700",
};

/** Style string for a cell. `stale` is applied last so a value known not to
 *  be current always looks stale, whatever else it is (§6.3 makes this
 *  mandatory). */
export function styleFor(attrs: Attr[]): string {
  const ordered = [...attrs].sort((a, b) => Number(a === "stale") - Number(b === "stale"));
  return ordered.map((a) => STYLES[a] ?? "").filter(Boolean).join(";");
}

/** A table pane as real HTML.
 *
 *  MDL, FLDS and SPTR are tables, and a pane type with no drawing renders
 *  as an empty box: the data arrives, conformance passes (it asserts
 *  region and binding, never pixels) and the user sees nothing. */
function renderTable(pane: LayoutPane, height: number, width: number): string {
  const rows = pane.data.map((row) => row.map((cell) => (cell === null ? "" : String(cell))));
  const body = rows
    .map((row) => `<tr>${row.map((cell) => `<td>${escapeHtml(cell)}</td>`).join("")}</tr>`)
    .join("");
  return (
    `<div class="pane pane-table" style="grid-row:span ${height};grid-column:span ${width}">`
    + `<table>${body}</table></div>`
  );
}

/** Render a pane. The desktop draws a real chart where the TUI draws a
 *  sparkline — §6.1 permits exactly this difference, which is why
 *  conformance asserts region/type/binding and never pixels. */
export function renderPane(pane: LayoutPane): string {
  const [, , height, width] = pane.region;
  if (pane.type === "table_scroll") return renderTable(pane, height, width);
  const points = pane.data
    .map((row) => row[row.length - 1])
    .filter((v): v is number => typeof v === "number");
  if (points.length < 2) {
    return `<div class="pane" style="grid-row:span ${height};grid-column:span ${width}">`
      + `<span class="pane-label">[${pane.type}:${pane.binding}]</span></div>`;
  }
  const lo = Math.min(...points);
  const hi = Math.max(...points);
  const span = hi - lo || 1;
  const path = points
    .map((v, i) => `${(i / (points.length - 1)) * 100},${100 - ((v - lo) / span) * 100}`)
    .join(" ");
  return (
    `<div class="pane" style="grid-row:span ${height};grid-column:span ${width}">`
    + `<span class="pane-label">[${pane.type}:${pane.binding}]</span>`
    + `<svg viewBox="0 0 100 100" preserveAspectRatio="none" class="pane-chart">`
    + `<polyline points="${path}" fill="none" stroke="#00afd7" stroke-width="1.5"/></svg></div>`
  );
}

function escapeHtml(text: string): string {
  return text.replace(/[&<>"']/g, (c) =>
    ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[c] as string);
}

// ---------------------------------------------------------------------
// CNVS — a whole workspace (spec §5.3)
// ---------------------------------------------------------------------

export interface CanvasPlacement {
  x: number; y: number; width: number; height: number; display: number;
}

export interface CanvasComponentTree {
  id: string;
  screen: string;
  channel: string | null;
  placement: CanvasPlacement | null;
  tree: LayoutTree;
}

/** Smallest placement with a visible interior. Mirrors MIN_PLACEMENT in
 *  treble/render/canvas.py; a component drawn with no room inside its frame
 *  is indistinguishable from one that failed to resolve. */
export const MIN_PLACEMENT = 3;

/** §6.3 tokens are for cells; a colour group is the *component's* channel and
 *  is drawn on its frame, because a link the user cannot see is a link they
 *  cannot verify. */
const CHANNEL_COLOURS: Record<string, string> = {
  red: "#d70000", blue: "#0087d7", green: "#00af5f",
  yellow: "#d7d700", orange: "#d78700", purple: "#af5fd7",
};

function onDisplay(components: CanvasComponentTree[], display: number): CanvasComponentTree[] {
  return components
    .filter((c) => c.placement !== null && c.placement.display === display)
    .sort((a, b) =>
      a.placement!.y - b.placement!.y
      || a.placement!.x - b.placement!.x
      || (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
}

function padEnd(text: string, length: number, fill: string): string {
  let out = text;
  while (out.length < length) out += fill;
  return out;
}

/** The plain-text projection of a workspace, character-identical to
 *  `canvas_text_snapshot` in treble/render/canvas.py. This is what the
 *  conformance suite compares (I6). */
export function canvasTextSnapshot(
  components: CanvasComponentTree[], display = 0,
): string {
  const placed = onDisplay(components, display);
  const height = placed.reduce((m, c) => Math.max(m, c.placement!.y + c.placement!.height), 0);
  const width = placed.reduce((m, c) => Math.max(m, c.placement!.x + c.placement!.width), 0);
  const grid: string[][] = Array.from({ length: height }, () => Array<string>(width).fill(" "));

  const write = (row: number, col: number, text: string): void => {
    for (let i = 0; i < text.length; i += 1) {
      if (row >= 0 && row < height && col + i >= 0 && col + i < width) grid[row][col + i] = text[i];
    }
  };

  for (const component of placed) {
    const p = component.placement!;
    if (p.width < MIN_PLACEMENT || p.height < MIN_PLACEMENT) {
      throw new Error(
        `component ${JSON.stringify(component.id)} is placed ${p.width}x${p.height}, which is `
        + `too small to draw anything inside a frame (minimum ${MIN_PLACEMENT}x${MIN_PLACEMENT}). `
        + "A component with no visible interior is indistinguishable from one that failed to resolve",
      );
    }
    const innerWidth = p.width - 2;
    const innerHeight = p.height - 2;
    const lines = textSnapshot(component.tree).replace(/\n+$/, "").split("\n");
    const clipped = lines.length > innerHeight || lines.some((l) => l.length > innerWidth);

    let title = ` ${component.screen}`;
    if (component.channel !== null) title += ` ${component.channel}`;
    if (clipped) title += " CLIP";
    title += " ";
    write(p.y, p.x, `+${padEnd(title.slice(0, innerWidth), innerWidth, "-")}+`);
    write(p.y + p.height - 1, p.x, `+${"-".repeat(innerWidth)}+`);
    for (let i = 0; i < innerHeight; i += 1) {
      const body = i < lines.length ? lines[i] : "";
      write(p.y + 1 + i, p.x, `|${padEnd(body.slice(0, innerWidth), innerWidth, " ")}|`);
    }
  }

  const rendered = grid.map((line) => line.join("").replace(/\s+$/, ""));
  const trailer: string[] = [];
  for (const component of [...components].sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0))) {
    if (component.placement === null) {
      trailer.push(
        `unplaced ${component.screen} ${component.channel ?? "unlinked"} id=${component.id}`);
    }
  }
  const elsewhere = components.filter(
    (c) => c.placement !== null && c.placement.display !== display).length;
  if (elsewhere > 0) {
    trailer.push(`display ${display} shown; ${elsewhere} component(s) on other displays`);
  }
  return [...rendered, ...trailer].join("\n").replace(/\n+$/, "") + "\n";
}

/** The styled workspace the desktop window displays: every component's own
 *  `renderHtml` output, absolutely positioned in workspace cells. */
export function renderCanvasHtml(components: CanvasComponentTree[], display = 0): string {
  const placed = onDisplay(components, display);
  const height = placed.reduce((m, c) => Math.max(m, c.placement!.y + c.placement!.height), 0);
  const width = placed.reduce((m, c) => Math.max(m, c.placement!.x + c.placement!.width), 0);

  const frames = placed.map((component) => {
    const p = component.placement!;
    const colour = component.channel ? CHANNEL_COLOURS[component.channel] ?? "#767676" : "#3a3a3a";
    const label = component.channel
      ? `${component.screen} · ${component.channel}` : component.screen;
    return (
      `<div class="canvas-component" data-component="${escapeHtml(component.id)}"`
      + ` style="grid-row:${p.y + 1}/span ${p.height};grid-column:${p.x + 1}/span ${p.width};`
      + `border:1px solid ${colour}">`
      + `<div class="canvas-title" style="color:${colour}">${escapeHtml(label)}</div>`
      + renderHtml(component.tree)
      + "</div>"
    );
  }).join("");

  // Unplaced components are listed, never dropped: a workspace that rendered
  // four of its six components would look like a working layout.
  const unplaced = components.filter((c) => c.placement === null);
  const note = unplaced.length === 0 ? "" : (
    `<div class="canvas-unplaced">${unplaced.length} component(s) with no placement: `
    + escapeHtml(unplaced.map((c) => `${c.screen} (${c.id})`).join(", "))
    + "</div>"
  );
  return (
    `<div class="canvas" style="display:grid;grid-template-columns:repeat(${width},1ch);`
    + `grid-template-rows:repeat(${height},1.2em)">${frames}</div>${note}`
  );
}

/** The plain-text projection, character-identical to every other renderer.
 *  This is what the conformance suite compares. */
export function textSnapshot(tree: LayoutTree): string {
  const [rows, cols] = tree.grid;
  const grid: string[][] = Array.from({ length: rows }, () => Array<string>(cols).fill(" "));

  const place = (row: number, col: number, text: string): void => {
    for (let i = 0; i < text.length; i += 1) {
      if (row >= 0 && row < rows && col + i >= 0 && col + i < cols) grid[row][col + i] = text[i];
    }
  };

  for (const cell of [...tree.cells].sort((a, b) => a.at[0] - b.at[0] || a.at[1] - b.at[1])) {
    place(cell.at[0], cell.at[1], cell.text);
  }

  // Panes use the renderer-neutral shape: conformance is about layout, not
  // the medium's drawing (CLAUDE.md §4).
  for (const pane of tree.panes) {
    const [row, col, height, width] = pane.region;
    const horizontal = width >= 2 ? `+${"-".repeat(width - 2)}+` : "+";
    place(row, col, horizontal);
    const label = `[${pane.type}:${pane.binding}]`.slice(0, Math.max(width - 2, 0));
    for (let i = 1; i < height - 1; i += 1) {
      place(row + i, col, `|${(i === 1 ? label : "").padEnd(Math.max(width - 2, 0))}|`);
    }
    if (height > 1) place(row + height - 1, col, horizontal);
  }

  return grid.map((line) => line.join("").replace(/\s+$/, "")).join("\n").replace(/\n+$/, "") + "\n";
}

/** The styled HTML the desktop window displays. */
export function renderHtml(tree: LayoutTree): string {
  const [rows, cols] = tree.grid;
  const lines: string[][] = Array.from({ length: rows }, () => Array<string>(cols).fill(" "));
  const styled: string[] = [];

  for (const cell of [...tree.cells].sort((a, b) => a.at[0] - b.at[0] || a.at[1] - b.at[1])) {
    const [row, col] = cell.at;
    const style = styleFor(cell.attrs);
    const attrs = cell.link ? ` data-command="${escapeHtml(cell.link)}"` : "";
    const trace = cell.provenance ? ` data-provenance="${escapeHtml(cell.provenance)}"` : "";
    styled.push(
      `<span class="cell" style="grid-row:${row + 1};grid-column:${col + 1}/span ${cell.text.length};${style}"${attrs}${trace}>`
      + `${escapeHtml(cell.text)}</span>`,
    );
    for (let i = 0; i < cell.text.length; i += 1) {
      if (col + i < cols) lines[row][col + i] = cell.text[i];
    }
  }

  const panes = tree.panes.map(renderPane).join("");
  return `<div class="grid" style="grid-template-columns:repeat(${cols},1ch);grid-template-rows:repeat(${rows},1.2em)">${styled.join("")}${panes}</div>`;
}

File: treble/render/tui/renderer.py
"""TUI renderer — CellBuffer to styled terminal output (spec §4, §6.1).

The renderer's only input is a `CellBuffer`. It knows nothing about
storage, analytics or where values came from, which is what makes the same
screen definition drive this and the desktop client (I6).

**Conformance.** This renderer must reproduce the same abstract layout tree
and text snapshot as every other renderer, from the same definition and the
same frozen TAPI responses (CLAUDE.md §4). It is registered with the
conformance suite so that is checked rather than assumed.

Graphical panes are the one legitimate difference (§6.1): where the desktop
client draws a WebGL chart, the TUI draws a braille sparkline. The suite
asserts the pane's region, type and data binding — never its pixels.
"""

from __future__ import annotations

from rich.text import Text

from treble.render.canvas import (
    Canvas,
    canvas_layout_tree,
    canvas_text_snapshot,
    components_on_display,
    frame_lines,
    normalise_content,
)
from treble.render.contract.buffer import (
    CellBuffer,
    ResolvedPane,
    layout_tree,
    pane_placeholder,
)
from treble.render.contract.schema import PaneType
from treble.render.tui.theme import DEFAULT_THEME, Theme

#: Braille dot rows, low to high. Eight levels in a single character cell,
#: which is what makes a text sparkline legible at terminal density.
_BRAILLE = "⣀⣄⣤⣦⣶⣷⣿"


def sparkline(values: list[float], width: int) -> str:
    """A braille sparkline — the TUI's answer to a chart pane.

    Returns an empty string for fewer than two points rather than drawing a
    flat line, which would imply data that is not there.
    """
    points = [v for v in values if v is not None]
    if len(points) < 2 or width < 1:
        return ""
    lo, hi = min(points), max(points)
    span = hi - lo
    # Resample to the available width.
    step = max(1, len(points) // width)
    sampled = points[::step][:width]
    if span == 0:
        return _BRAILLE[len(_BRAILLE) // 2] * len(sampled)
    return "".join(
        _BRAILLE[min(len(_BRAILLE) - 1, int((v - lo) / span * (len(_BRAILLE) - 1)))]
        for v in sampled
    )


def table_lines(pane: ResolvedPane, *, height: int, width: int) -> list[str]:
    """A table pane drawn as aligned columns.

    Needed because MDL, FLDS and SPTR are tables, and a pane type with no
    drawing renders as an empty box: the data resolves, conformance passes
    (it asserts region and binding, never pixels) and the user sees
    nothing. Truncation is always announced for the same reason — a table
    silently cut off is a wrong display, not a tidy one.
    """
    rows = [["" if cell is None else str(cell) for cell in row] for row in pane.data]
    if not rows:
        return [f"[{pane.binding}] no rows"[:width]]

    columns = max(len(row) for row in rows)
    widths = [max((len(row[i]) for row in rows if i < len(row)), default=0) for i in range(columns)]

    gap = 2
    # Shrink the widest column repeatedly rather than scaling everything:
    # it keeps short identifier columns intact and truncates prose.
    while sum(widths) + gap * (columns - 1) > width and max(widths) > 4:
        widths[widths.index(max(widths))] -= 1

    visible = rows
    footer = ""
    if len(rows) > height:
        visible = rows[: height - 1]
        footer = f"... {len(rows) - len(visible)} more rows"

    lines = [
        (" " * gap)
        .join(
            (row[i] if i < len(row) else "")[: widths[i]].ljust(widths[i]) for i in range(columns)
        )
        .rstrip()[:width]
        for row in visible
    ]
    if footer:
        lines.append(footer[:width])
    return lines


def render_pane(pane: ResolvedPane) -> list[str]:
    """A pane as text lines. Region and binding are honoured exactly; the
    *drawing* is the TUI's own (§6.1)."""
    height, width = pane.region.height, pane.region.width
    lines: list[str] = []
    if pane.pane_type is PaneType.TIMESERIES and pane.data:
        numeric = [float(row[-1]) for row in pane.data if row and isinstance(row[-1], int | float)]
        spark = sparkline(numeric, max(width - 2, 1))
        lines.append(f"[{pane.pane_type.value}:{pane.binding}]"[:width])
        if spark:
            lines.append(spark)
    elif pane.pane_type is PaneType.TABLE_SCROLL:
        lines.extend(table_lines(pane, height=height, width=width))
    else:
        lines.append(f"[{pane.pane_type.value}:{pane.binding}]"[:width])
    while len(lines) < height:
        lines.append("")
    return lines[:height]


def render_text(buffer: CellBuffer) -> Text:
    """The buffer as styled Rich text, one line per grid row."""
    return render_styled(buffer, DEFAULT_THEME)


def styled_grid(
    buffer: CellBuffer, theme: Theme, *, draw_panes: bool = True
) -> tuple[list[list[str]], list[list[str]]]:
    """(characters, styles) for a buffer, as parallel row-major grids.

    Split out from :func:`render_styled` so the canvas compositor can place a
    component's characters *and* its styling without a second implementation
    of either. A style grid rebuilt independently would be a second source of
    truth for what a cell looks like.
    """
    grid: list[list[str]] = [[" "] * buffer.cols for _ in range(buffer.rows)]
    styles: list[list[str]] = [[""] * buffer.cols for _ in range(buffer.rows)]

    def place(row: int, col: int, text: str, style: str) -> None:
        for offset, char in enumerate(text):
            if 0 <= row < buffer.rows and 0 <= col + offset < buffer.cols:
                grid[row][col + offset] = char
                styles[row][col + offset] = style

    for cell in sorted(buffer.cells, key=lambda c: (c.row, c.col)):
        place(cell.row, cell.col, cell.text, theme.style_for(cell.attrs))

    for pane in buffer.panes:
        lines = render_pane(pane) if draw_panes else pane_placeholder(pane)
        for offset, line in enumerate(lines):
            place(pane.region.row + offset, pane.region.col, line, "")

    return grid, styles


def _to_text(grid: list[list[str]], styles: list[list[str]]) -> Text:
    """Collapse parallel character/style grids into Rich runs."""
    out = Text()
    rows, cols = len(grid), len(grid[0]) if grid else 0
    for row_index in range(rows):
        col_index = 0
        while col_index < cols:
            style = styles[row_index][col_index]
            run = grid[row_index][col_index]
            col_index += 1
            while col_index < cols and styles[row_index][col_index] == style:
                run += grid[row_index][col_index]
                col_index += 1
            out.append(run.rstrip() if col_index >= cols else run, style=style or None)
        if row_index < rows - 1:
            out.append("\n")
    return out


def render_styled(buffer: CellBuffer, theme: Theme, *, draw_panes: bool = True) -> Text:
    """Render with an explicit theme (§6.3: semantics survive the palette).

    ``draw_panes=False`` substitutes the renderer-neutral pane shape used
    for conformance; display always draws the real sparkline.
    """
    return _to_text(*styled_grid(buffer, theme, draw_panes=draw_panes))


def render_canvas_styled(
    canvas: Canvas,
    buffers: dict[str, CellBuffer],
    theme: Theme,
    *,
    display: int = 0,
    draw_panes: bool = True,
) -> Text:
    """A whole workspace as styled terminal output (spec §5.3).

    The frame geometry comes from :func:`~treble.render.canvas.frame_lines`,
    the same function the text projection uses, so this cannot drift from it.
    What the TUI adds is the *styling*: each component's cells keep their
    semantic attributes, and the frame is drawn in its colour group so a link
    is visible on the terminal exactly as it is on the desktop.

    A terminal has one grid and no windows, so `display` selects which
    display's components to draw rather than showing all of them at once —
    overlaying two displays' placements would collide arbitrary components.
    """
    on_display = components_on_display(canvas, buffers, display)
    height = max((p.y + p.height for _, p, _ in on_display), default=0)
    width = max((p.x + p.width for _, p, _ in on_display), default=0)
    grid: list[list[str]] = [[" "] * width for _ in range(height)]
    styles: list[list[str]] = [[""] * width for _ in range(height)]

    for component, placement, buffer in on_display:
        component_grid, component_styles = styled_grid(buffer, theme, draw_panes=draw_panes)
        content = normalise_content(["".join(row) for row in component_grid])
        frame = frame_lines(component, placement, content)
        frame_style = theme.style_for_channel(
            component.channel.value if component.channel else None
        )
        for row_offset, line in enumerate(frame):
            row = placement.y + row_offset
            if not 0 <= row < height:
                continue
            for col_offset, char in enumerate(line):
                col = placement.x + col_offset
                if not 0 <= col < width:
                    continue
                grid[row][col] = char
                # Interior characters keep the component's own styling; the
                # border and title carry the colour group. `frame_lines` has
                # already clipped and padded, so the two agree on which
                # characters exist by construction rather than by luck.
                inner_row = row_offset - 1
                inner_col = col_offset - 1
                interior = (
                    0 <= inner_row < len(component_styles)
                    and 0 <= inner_col < len(component_styles[inner_row])
                    and 0 < row_offset < placement.height - 1
                    and 0 < col_offset < placement.width - 1
                )
                styles[row][col] = (
                    component_styles[inner_row][inner_col] if interior else frame_style
                )

    return _to_text(grid, styles)


def canvas_conformance_artifacts(
    canvas: Canvas, buffers: dict[str, CellBuffer], *, display: int = 0
) -> tuple[str, str]:
    """(canvas layout tree, canvas text snapshot) from the TUI's own pipeline.

    The snapshot is taken from the styled workspace render with styling
    stripped, for the same reason the single-screen version is: it proves
    what this renderer actually draws, not that a second code path agrees.
    """
    styled = render_canvas_styled(canvas, buffers, DEFAULT_THEME, display=display, draw_panes=False)
    lines = [line.rstrip() for line in styled.plain.split("\n")]
    trailer = canvas_text_snapshot(canvas, buffers, display=display).rstrip("\n").split("\n")
    # The trailer (unplaced components, other displays) is workspace
    # bookkeeping rather than something drawn on the grid, so it is appended
    # from the shared projection instead of re-derived here.
    body_height = len(lines)
    text = "\n".join([*lines, *trailer[body_height:]]).rstrip("\n") + "\n"
    return canvas_layout_tree(canvas, buffers), text


def conformance_artifacts(buffer: CellBuffer) -> tuple[str, str]:
    """(layout tree, text snapshot) as the conformance suite requires.

    The text snapshot is taken from the *styled* render with styling
    stripped, so this proves the TUI's own pipeline produces the agreed
    output — not that a second code path happens to agree.
    """
    styled = render_styled(buffer, DEFAULT_THEME, draw_panes=False)
    text = "\n".join(line.rstrip() for line in styled.plain.split("\n"))
    return layout_tree(buffer), text.rstrip("\n") + "\n"
